### PR TITLE
feat: parse YAML/JSON/JS config file

### PR DESCRIPTION
**Pull request by honzahommer**
**Type:** feature

**Description:**

Allow to use JavaScript exported module config file.

```sh
$ cat ~/config.js
module.exports = require('config');

$ verdaccio --config ~/config.js
```

### CODE REVIEW

File: src/lib/utils.js [hunk 1]
```diff
@@ -380,8 +380,17 @@ export const ErrorCode = {
   },
 };
 
-export function parseConfigFile(configPath: string): Object {
-  return YAML.safeLoad(fs.readFileSync(configPath, CHARACTER_ENCODING.UTF8));
+function parseConfigFile(configPath: string) {
```
Comment (juanpicado) [inline src/lib/utils.js]: `export` is missing.

Comment (honzahommer) [reply src/lib/utils.js]: My mistake, fixed.

Comment (juanpicado) [reply src/lib/utils.js]:
```
/home/circleci/verdaccio/src/lib/utils.js
  386:10  error  Insert `·`  prettier/prettier
  389:12  error  Insert `·`  prettier/prettier
```
still prettiers yells, just that little thing please :-)

File: test/unit/api/config.spec.js [hunk 2]
```diff
@@ -81,6 +81,13 @@ describe('Config file', () => {
       expect(config.auth.htpasswd.file).toBe('./htpasswd');
       checkDefaultConfPackages(config);
     });
+
+    test('parse full.json', () => {
+      const config = new Config(parseConfigFile(resolveConf('full').replace(/.yaml$/, '.json')));
+      checkDefaultUplink(config);
+      expect(config.storage).toBe('./storage');
+      checkDefaultConfPackages(config);
```
Comment (juanpicado) [inline test/unit/api/config.spec.js]: I'd add the scenario where the JSON has bad formating.

**Comment by alexjurkiewicz** [comment]
YAML is a superset of JSON, so this is not necessary. You can just pass in a JSON file now:


```
> a
'{\n  "storage": "./storage",\n  "plugins": "./plugins",\n  "web": {\n    "title": "Verdaccio"\n  },\n  "auth": {\n    "htpasswd": {\n      "file": "./htpasswd"\n    }\n  },\n  "middlewares": {\n    "audit": {\n      "enabled": true\n    }\n  },\n  "uplinks": {\n    "npmjs": {\n      "url": "https://registry.npmjs.org/"\n    }\n  },\n  "packages": {\n    "@*/*": {\n      "access": "$all",\n      "publish": "$authenticated",\n      "proxy": "npmjs"\n    },\n    "**": {\n      "access": "$all",\n      "publish": "$authenticated",\n      "proxy": "npmjs"\n    }\n  },\n  "server": {\n    "keepAliveTimeout": 60\n  },\n  "logs": [\n    {\n      "type": "stdout",\n      "format": "pretty",\n      "level": "http"\n    }\n  ]\n}'
> YAML.parse(a)
{ storage: './storage',
  plugins: './plugins',
  web: { title: 'Verdaccio' },
  auth: { htpasswd: { file: './htpasswd' } },
  middlewares: { audit: { enabled: true } },
  uplinks: { npmjs: { url: 'https://registry.npmjs.org/' } },
  packages:
   { '@*/*': { access: '$all', publish: '$authenticated', proxy: 'npmjs' },
     '**': { access: '$all', publish: '$authenticated', proxy: 'npmjs' } },
  server: { keepAliveTimeout: 60 },
  logs: [ { type: 'stdout', format: 'pretty', level: 'http' } ] }
```

File: src/lib/utils.js (hunk 1 above)
Comment (honzahommer) [reply src/lib/utils.js]: fixed

**Comment by honzahommer** [comment]
@alexjurkiewicz This is for Node.js Module Exports.

**Comment by iDaN5x** [comment]
@honzahommer could you elaborate? 
I agree with @alexjurkiewicz 's statement about the PR not being necessary  to support the use-case.

**Comment by honzahommer** [comment]
@iDaN5x This is my use case (maybe I am wrong, but at this moment I am not able to do this):

```
> ~/config.js
module.exports = require('config');

$ verdaccio --config ~/config.js
```

**Comment by alexjurkiewicz** [comment]
OK. That's not a JSON config file, but a Javascript file. You want Verdaccio to execute custom code to load your config.

One problem with this approach is that your code must generate a JS object which exactly matches the config format that Verdaccio expects. When Verdaccio loads a YAML file, it can modify the representation before using it as a config file (example use case: config file version migrations).

How would you feel about this workflow instead?

```bash
cat <<EOF >gen-config.js
const config = require('config');
JSON.stringify(obj)
fs.writeFileSync('verdaccio.json', config);
EOF

node gen-config.js
verdaccio --config verdaccio.json
```

In short, you modify your code to write the config object to disk as JSON. Then you can load it as a normal config file. This would work today (I believe). Would that work for you?

**Comment by honzahommer** [comment]
@alexjurkiewicz I know, and I am doing it as you suggest at this moment. I think this feature should be useful for many users to allow pass JavaScript file path as config instead of converting into JSON and writing to disk.

File: conf/full.json [hunk 3]
```diff
@@ -0,0 +1,44 @@
+{
+  "storage": "./storage",
```
Comment (juanpicado) [inline conf/full.json]: I think this should be removed completely, as you can see https://github.com/verdaccio/verdaccio/blob/master/conf/full.yaml it's empty due I was used as reference and force us to maintain multiple config example files.

File: test/unit/api/config.spec.js [hunk 4]
```diff
@@ -81,6 +81,13 @@ describe('Config file', () => {
       expect(config.auth.htpasswd.file).toBe('./htpasswd');
       checkDefaultConfPackages(config);
     });
+
+    test('parse full.json', () => {
+      const config = new Config(parseConfigFile(resolveConf('full').replace(/.yaml$/, '.json')));
```
Comment (juanpicado) [inline test/unit/api/config.spec.js]: Since `yaml` is the official config, the other blocks are ok, but this one, would need a special treatment, a mock file within the `partials`. 

Something else to update in this ticket needs also updates documentation https://github.com/verdaccio/website/blob/master/docs/cli.md defining the JSON format is also allowed.

**Comment by juanpicado** [comment]
```
verdaccio --config verdaccio.json
```
I'm fine with this, I think last year I was asked have JSON format as input is handy, since allows more flexible configuration, as, fetching config from and endpoint or something like that. It will bring new ideas of run verdaccio fo sure.

**Comment by honzahommer** [comment]
@juanpicado I'm not sure that I am understand to your last comment about `mock file within the partials`. Could you please provide more informations?

File: test/unit/api/config.spec.js (hunk 4 above)
Comment (honzahommer) [reply test/unit/api/config.spec.js]: @juanpicado I'm not sure that I am understand to your last comment about mock file within the partials - could you please write one "example" test, then I can write the rest.

Comment (juanpicado) [reply test/unit/api/config.spec.js]: If you see in this location, there is a folder called `yaml`, it is where I started to host all examples used for unit testing, I think we need one called `json` and restore the unit test your have removed. For instance:
```
const parseJSONConfigurationFile = (name) => {
    return path.join(__dirname, `../partials/config/json/${name}.json`);
  };

test('parse valid example.json', () => {
      const config = new Config(parseConfigurationFile('example1'));
      checkDefaultUplink(config);
      expect(config.storage).toBe('./storage');
      checkDefaultConfPackages(config);
    });

test('parse invalid example.json', () => {
      expect(function ( ) {
           const config = new Config(parseConfigurationFile('example1'));
       }).toThrow(Error('should fail '));
    });

```

I just made that without run it on my terminal :-) , but that's the idea, to cover those scenarios. It will help if someone changes something in your code and stop to work as is intended.

https://github.com/verdaccio/verdaccio/tree/master/test/unit/partials/config

**Comment by juanpicado** [comment]
I also would add a comment here => https://github.com/verdaccio/website/blob/master/docs/cli.md#commands about you actually can use JSON format, but, I'd point out only works with absolute paths!

**Comment by iDaN5x** [comment]
@honzahommer this is the code flow that causes the test to fail (correctly):
1. Invoking `fs.loadFileSync` on the file path returns the string `module.exports = {;`
2. This string is then parsed by `YAML.safeLoad` as a plain string, resulting with the string staying as-is.
3. The `parseConfigFile` then returns the string and the function execution completes successfully.
4. Seeing that the function didn't throw, the test passes.

The meaning of this is that the implementation can't be trusted to work correctly.
Looking at `js-yaml` library, it doesn't seem theres a function that loads YAML directly from file.

There are two ways of dealing with this issue:
We can either replace `js-yaml` with some other library who does support file-loading.
Or, we can check if the file extension is `.js` and behave accordingly.

My preferable solution is checking the file extension.
However, I'm still not seeing a concrete use-case which supports this feature...

**Comment by honzahommer** [comment]
**PoC**

There is possibility to use require.extensions to handle loading YAML file. This option is officially deprecated (https://nodejs.org/api/all.html#modules_require_extensions), but works fine. This also fixed issues with non-absolute paths to config file as you mentioned above.

```js
/**
 * Parse and returns config file.
 * @param {String} configPath
 * @return {Object}
 * @throws Will throw an error if the configPath is not resolved.
 */
export function parseConfigFile(configPath: string) {
  const _require = _.cloneDeep(require);

  _require.extensions['.yml'] = _require.extensions['.yaml'] = function (module, filename) {
    module.exports = YAML.safeLoad(fs.readFileSync(filename, CHARACTER_ENCODING.UTF8));

    if (isObject(module.exports) === false) {
      throw Error(APP_ERROR.CONFIG_NOT_VALID);
    }

    return module;
  };

  try {
    return _require(configPath);
  } catch (e) {
    e.message = e.code === 'MODULE_NOT_FOUND' ? CONFIG_NOT_FOUND : APP_ERROR.CONFIG_NOT_VALID;

    throw Error(e);
  }
};
```

**Comment by iDaN5x** [comment]
@honzahommer I don't like this solution for two reasons:
A. require extensions are deperecated, as you mentioned. Even if they work now, they'll most definitely be removed in the near future.
B. Even if they weren't deprecated, they make intrusive changes to the runtime itself, that are hard to debug in case of a bug or unintended behavior. This is one of the reasons they were deprecated in the first place

But let me ask you once again - do you have a concrete use-case requiring this feature? Can you give examples of cases where the configuration needs to be generated on boot?

I guess the inspiration for this feature comes from different NodeJS __frameworks__ such as Webpack (packaging) or Jester (testing). Please remember that Verdaccio is a __complete system__ and not a framework for other systems.

**Comment by juanpicado** [comment]
@iDaN5x I have a use case => https://github.com/verdaccio/ui/blob/4.x-master/tools/verdaccio.js#L7

I must import `js-yaml` in each project and is quite annoying. I'd prefer to have something native, JSON preferably.

**Comment by juanpicado** [comment]
@honzahommer sorry the delay. The PR works pretty fine, tests were failing due to my fault on the master branch, my bad.

Now, I need your help updating https://github.com/verdaccio/website/blob/master/docs/cli.md#commands providing enough context for users want to use your new feature. Could you PR there?

**Comment by juanpicado** [comment]
Ok, actually the test fails because

```
 Config Utilities › JavaScript › parse invalid.js

    expect(received).toThrow(expected)

    Expected pattern: /Error/

    Received function did not throw

      300 |       expect(function ( ) {
      301 |         parseConfigFile(parseConfigurationFile('invalid.js'));
    > 302 |       }).toThrow(/Error/);
```

Screen Shot 2019-04-26 at 10 36 25 PM

If the JS is invalid, the `return YAML.safeLoad(fs.readFileSync(configPath, CHARACTER_ENCODING.UTF8));` actually is able to parse it and return a string, as my attachment above.

So, unfortunately, we cannot rely on the current implementation. I think @iDaN5x was right and we need to check the extension.

**Comment by honzahommer** [comment]
> @honzahommer sorry the delay. ~The PR works pretty fine, thanks for it, tests were failing due to my fault on the master branch, my bad.~
> 
> Now, I need your help updating https://github.com/verdaccio/website/blob/master/docs/cli.md#commands providing enough context for users want to use your new feature. Could you PR there?

See https://github.com/verdaccio/website/pull/99.

**Comment by honzahommer** [comment]
> An invalid JS is being parsed by `yaml` library.

https://github.com/verdaccio/verdaccio/blob/44105d151dce5dda069071350150c2bd507df885/src/lib/utils.js#L396

**Comment by juanpicado** [comment]
@honzahommer good :-) I'll test it deeply after v4 release.

**Comment by juanpicado** [comment]
@honzahommer could you resolve the conflicts?

**Comment by honzahommer** [comment]
> @honzahommer could you resolve the conflicts?

I've merged with 4.1.x, conflicts are resolved.
